runtime/src/main/kotlin/kotlin/native/Text.kt

```diff
@@ -9,52 +9,103 @@ package kotlin.native
 /**
  * Converts an UTF-8 array into a [String]. Replaces invalid input sequences with a default character.
  */
+@Deprecated("Use decodeToString or decodeZeroTerminatedToString instead")
 public fun ByteArray.stringFromUtf8() : String {
+    @Suppress("DEPRECATION")
     return this.stringFromUtf8(0, this.size)
 }
 
 /**
  * Converts an UTF-8 array into a [String]. Replaces invalid input sequences with a default character.
  */
+@Deprecated("Use decodeToString or decodeZeroTerminatedToString instead")
 public fun ByteArray.stringFromUtf8(start: Int = 0, size: Int = this.size) : String {
-    checkBoundsIndexes(start, start + size, this.size)
-    return stringFromUtf8Impl(start, realSize(this, size))
+    val endIndex = start + size
+    checkBoundsIndexes(start, endIndex, this.size)
+
+    val realSize = realEndIndex(this, start, endIndex) - start
+    return unsafeStringFromUtf8Impl(start, realSize)
 }
 
-@SymbolName("Kotlin_ByteArray_stringFromUtf8")
-internal external fun ByteArray.stringFromUtf8Impl(start: Int, size: Int) : String
+/**
+ * Decodes a string from the bytes in UTF-8 encoding in this array.
+ * Bytes following the first occurrence of `0` byte, if it occurs, are not decoded.
+ *
+ * Malformed byte sequences are replaced by the replacement char `\uFFFD`.
+ */
+@SinceKotlin("1.3")
+public fun ByteArray.decodeZeroTerminatedToString() : String {
+    return unsafeStringFromUtf8Impl(0, realEndIndex(this, 0, this.size))
+}
 
-private fun realSize(byteArray: ByteArray, size: Int): Int {
-    var realSize = 0
-    while (realSize < size && byteArray[realSize] != 0.toByte()) {
-        realSize++
+/**
+ * Decodes a string from the bytes in UTF-8 encoding in this array or its subrange.
+ * Bytes following the first occurrence of `0` byte, if it occurs, are not decoded.
+ *
+ * @param startIndex the beginning (inclusive) of the subrange to decode, 0 by default.
+ * @param endIndex the end (exclusive) of the subrange to decode, size of this array by default.
+ * @param throwOnInvalidSequence specifies whether to throw an exception on malformed byte sequence or replace it by the replacement char `\uFFFD`.
+ *
+ * @throws IndexOutOfBoundsException if [startIndex] is less than zero or [endIndex] is greater than the size of this array.
+ * @throws IllegalArgumentException if [startIndex] is greater than [endIndex].
+ * @throws CharacterCodingException if the byte array contains malformed UTF-8 byte sequence and [throwOnInvalidSequence] is true.
+ */
+@SinceKotlin("1.3")
+public fun ByteArray.decodeZeroTerminatedToString(
```

Comment:
One of unfortunate things is that we use ByteArray here. Technically, anything, which can get a byte by index (i.e. interop pointers) shall be usable for such an API. @SvyatoslavScherbina